File: Wire-iOS/Sources/UserInterface/UserProfile/Views/ProfileFooterView.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "ProfileFooterView.h"

@import PureLayout;
#import "IconButton.h"
#import "NSLayoutConstraint+Helpers.h"
#import "Wire-Swift.h"



@implementation ProfileFooterView

- (id)init
{
    return [self initWithFrame:CGRectZero];
}

- (id)initWithFrame:(CGRect)frame
{
    if (self = [super initWithFrame:frame]) {
        [self createViews];
        [self setupConstraints];
    }
    return self;
}

- (void)createViews
{
    self.leftButton = [[IconButton alloc] initWithStyle:IconButtonStyleDefault];
    self.leftButton.translatesAutoresizingMaskIntoConstraints = NO;
    self.leftButton.accessibilityIdentifier = @"left_button";
    [self.leftButton setTitleColor:[UIColor wr_colorFromColorScheme:ColorSchemeColorTextForeground] forState:UIControlStateNormal];
    [self.leftButton setTitleColor:[UIColor wr_colorFromColorScheme:ColorSchemeColorTextDimmed] forState:UIControlStateHighlighted];
    [self.leftButton setTitleImageSpacing:16];
    self.leftButton.titleLabel.font = UIFont.smallLightFont;
    [self addSubview:self.leftButton];
    
    self.rightButton = [[IconButton alloc] initWithStyle:IconButtonStyleCircular];
    self.rightButton.translatesAutoresizingMaskIntoConstraints = NO;
    self.rightButton.accessibilityIdentifier = @"right_button";
    [self addSubview:self.rightButton];
}

- (void)setupConstraints
{
    [self.leftButton addConstraintForLeftMargin:16 relativeToView:self];
    [self.leftButton addConstraintForBottomMargin:12 relativeToView:self];
    [self.leftButton addConstraintForTopMargin:0 relativeToView:self];
    [self.leftButton autoSetDimension:ALDimensionHeight
                               toSize:32];
    
    [self.rightButton addConstraintForRightMargin:8 relativeToView:self];
    [self.rightButton addConstraintForBottomMargin:12 relativeToView:self];
    [self.rightButton addConstraintForTopMargin:0 relativeToView:self];
    [self.rightButton autoMatchDimension:ALDimensionWidth
                             toDimension:ALDimensionHeight
                                  ofView:self.rightButton];
    [self.rightButton autoSetDimension:ALDimensionWidth
                                toSize:32];
}

- (void)setIconTypeForLeftButton:(ZetaIconType)iconType
{
    self.leftButton.hidden = (iconType == ZetaIconTypeNone);
    if (iconType != ZetaIconTypeNone) {
        [self setIconType:iconType forButton:self.leftButton];
    }
}

- (void)setIconTypeForRightButton:(ZetaIconType)iconType
{
    self.rightButton.hidden = (iconType == ZetaIconTypeNone);
    if (iconType != ZetaIconTypeNone) {
        [self setIconType:iconType forButton:self.rightButton];
    }
}

- (void)setIconType:(ZetaIconType)iconType forButton:(IconButton *)button
{
    [button setIcon:iconType withSize:ZetaIconSizeTiny forState:UIControlStateNormal];
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Views/ProfileFooterView.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

// MARK: - ProfileFooterViewDelegate

protocol ProfileFooterViewDelegate: AnyObject {
    /// Called when the footer wants to perform a single action, from the left button.
    func footerView(_ footerView: ProfileFooterView, shouldPerformAction action: ProfileAction)

    /// Called when the footer wants to present the list of actions, from the right button.
    func footerView(_ footerView: ProfileFooterView, shouldPresentMenuWithActions actions: [ProfileAction])
}

// MARK: - ProfileFooterView

/// The footer of to use in the profile details screen.

final class ProfileFooterView: ConversationDetailFooterView {
    /// The object that will perform the actions on demand.
    weak var delegate: ProfileFooterViewDelegate?

    /// The action on the left button.
    var leftAction: ProfileAction?

    /// The actions hidden behind the ellipsis on the right.
    var rightActions: [ProfileAction]?

    // MARK: - Configuration

    override func setupButtons() {
        leftButton.accessibilityIdentifier = "left_button"
        rightButton.accessibilityIdentifier = "right_button"
        rightButton.accessibilityLabel = L10n.Localizable.Meta.Menu.accessibilityMoreOptionsButton
    }

    /// Configures the footer to display the specified actions.
    /// - parameter actions: The actions to display in the footer.

    func configure(with actions: [ProfileAction]) {
        // Separate the last and first actions
        var leftAction = actions.first
        var rightActions: [ProfileAction]

        if leftAction?.isEligibleForKeyAction == true {
            rightActions = Array(actions.dropFirst())
        } else {
            // If the first action is not eligible for key action, display
            // everything on the right
            leftAction = nil
            rightActions = actions
        }

        self.leftAction = leftAction
        self.rightActions = rightActions

        // Display the left action
        if let leftAction {
            leftButton.setTitle(leftAction.buttonText, for: .normal)
            leftIcon = leftAction.keyActionIcon
        }

        // Display or hide the right action ellipsis
        if rightActions.isEmpty {
            rightIcon = nil
            rightButton.isHidden = true
        } else {
            rightIcon = .ellipsis
            rightButton.isHidden = false
        }
    }

    // MARK: - Events

    override func leftButtonTapped(_: IconButton) {
        guard let leftAction else { return }
        delegate?.footerView(self, shouldPerformAction: leftAction)
    }

    override func rightButtonTapped(_: IconButton) {
        guard let rightActions else { return }
        delegate?.footerView(self, shouldPresentMenuWithActions: rightActions)
    }
}
